src/shared/machine-id-setup.c

```diff
@@ -40,15 +41,20 @@ static int acquire_machine_id_from_credential(sd_id128_t *ret) {
         if (r == 0) /* not found */
                 return -ENXIO;
 
-        r = sd_id128_from_string(buf, ret);
+        if (streq(buf, "firmware")) {
+                *ret_machine_id_from_firmware = true;
```

Comment:
hmm, so we have this rule in systemd that functions should initialize all return params on success, and leave them untouched on failure.

Here, you initialize only one of the two return params if we return success...

Hence, please initialize *ret_machine_id to SD_ID128_NULL here.